File: packages/shared/src/cognitive/effective-trading-policy.ts
/**
 * EffectiveTradingPolicy — TradingPolicy resuelta desde perfil activo (V1.30).
 * SoT para límites de encaje en scenario / priority; fallback moderate sin perfil.
 */

import type { PolicyTemplateId, TradingPolicyV1 } from "./trading-policy.js";
import { getTradingPolicyTemplate } from "./trading-policy-templates.js";

export type EffectiveTradingPolicyTemplateId = Extract<
  PolicyTemplateId,
  "conservative" | "moderate" | "aggressive_swing"
>;

function normalizeTemplateId(
  templateId: string | null | undefined,
): EffectiveTradingPolicyTemplateId {
  if (templateId === "conservative" || templateId === "aggressive_swing") {
    return templateId;
  }
  return "moderate";
}

/** Política operativa efectiva desde plantilla de perfil (≠ permiso de firma). */
export function resolveEffectiveTradingPolicy(
  templateId: string | null | undefined,
): TradingPolicyV1 {
  return getTradingPolicyTemplate(normalizeTemplateId(templateId));
}

export function effectiveMaxSectorExposurePct(
  templateId: string | null | undefined,
): number {
  return resolveEffectiveTradingPolicy(templateId).exposure
    .maxSectorExposurePct;
}

export function effectiveMaxPortfolioConcentrationPct(
  templateId: string | null | undefined,
): number {
  return resolveEffectiveTradingPolicy(templateId).exposure
    .maxPortfolioConcentrationPct;
}

/** Preview Config — límites de encaje (no es permiso). */
export function formatPortfolioFitPreview(policy: TradingPolicyV1): string {
  const e = policy.exposure;
  return `Encaja: max sector ${e.maxSectorExposurePct}% · max posición ${e.maxPortfolioConcentrationPct}% · ${e.maxOpenPositions} pos. abiertas`;
}

File: packages/py/analytics/src/bolsa_analytics/cognitive/effective_trading_policy.py
"""EffectiveTradingPolicy — límites de encaje desde plantilla de perfil (V1.30)."""

from __future__ import annotations

from bolsa_analytics.cognitive.trading_policy import TradingPolicy
from bolsa_analytics.cognitive.trading_policy_templates import get_policy_template

_EFFECTIVE_TEMPLATE_IDS = frozenset({"conservative", "moderate", "aggressive_swing"})


def resolve_effective_trading_policy(template_id: str | None) -> TradingPolicy:
    """Política operativa efectiva; fallback moderate sin perfil."""
    tid = template_id if template_id in _EFFECTIVE_TEMPLATE_IDS else "moderate"
    return get_policy_template(tid)


def effective_max_sector_exposure_pct(template_id: str | None) -> float:
    return resolve_effective_trading_policy(template_id).exposure.max_sector_exposure_pct


def format_portfolio_fit_preview(policy: TradingPolicy) -> str:
    e = policy.exposure
    return (
        f"Encaja: max sector {e.max_sector_exposure_pct:.0f}% · "
        f"max posición {e.max_portfolio_concentration_pct:.0f}% · "
        f"{e.max_open_positions} pos. abiertas"
    )
